YALB-615: Long Breadcrumbs

ModulesUnraveled opened the pull request:
## [YALB-615: Long Breadcrumbs](https://yaleits.atlassian.net/browse/YALB-615)

### Description of work
- Adds js/css to accommodate long breadcrumbs similar to how we're doing tabs

### Testing Link(s)
- [ ] Navigate to the [Component Story](https://deploy-preview-130--dev-component-library-twig.netlify.app/?path=/story/page-examples-standard-pages--basic)

### Functional Review Steps
- [ ] Verify you can install the component with the CLI
- [ ] Visit the page above, and verify the long breadcrumbs look and function at ALL screen sizes.
	- [ ] Also go to mobile, and expand the breadcrumbs to verify they work in either state

### Design Review
- [ ] Verify the designs match the [Figma Designs](https://www.figma.com/file/OVfmCW5s1gI7xnrM8ibX2y/UI-Kit-Alpha-%5BYaleSites%5D)
- [ ] Review any new [Percy builds](https://percy.io/95144ff9/component-library-twig)

### Accessibility Review
- [ ] Verify the component meets Accessibility requirements

ntay90 reviewed (changes requested):
The focus indicator does not appear for the breadcrumbs (or any other component).  In order for me  to proceed with an accessibility review, a focus indicator must be visible.

ModulesUnraveled commented:
@ntay90 I'm not sure what to say, we definitely have focus indicators... What browser/tech are you using? If you click on the page, and tab through does the indicator not show up on anything?

ntay90 commented:
@ModulesUnraveled I tested with Chrome, Safari, and Edge and had no luck; nothing on the canvas is producing a focus indicator.

https://user-images.githubusercontent.com/69398223/187462212-7a01de5c-9009-4dc8-93b3-11e22ce943f9.mov

ModulesUnraveled commented:
Dang! I should have looked at the deployed site before I commented. There was a build issue. I just rebuilt it, and everything looks correct to me now. (I also updated the demo link to point to the [deployed storybook](https://deploy-preview-130--dev-component-library-twig.netlify.app/?path=/story/page-examples-standard-pages--basic))

ntay90 commented:
Perfect! It works for me now. Looking at it, accessibility-wise, is it possible for the left/right buttons to receive keyboard focus/functionality? It's currently not possible for a keyboard user to view the entirety of the last breadcrumb.

callinmullaney reviewed (changes requested):
Left some code comments. I also have some functional questions/observations:
1. The completely collapsed version only shows the most recent parent as the trigger. This is in contrast to the expanded version including the current page as the last item. Seems that the trigger should be recent parent + page title OR leave it how it is now but also remove the page title from the mobile size expanded version.
2. When the completely collapsed version is triggered to expand it starts from Home (the beginning) and I have to scroll to the right in order to see the trigger text I just pressed. I think this should expand to the last item being visible and the left arrow being present.
3. Should the current page be focus-able? This is probably an edge-case but when I expand the breadcrumbs and try to tab over so it triggers the auto-scroll I can't see the whole page title. I have to then click the arrow for the breadcrumbs to fully scroll to the right.
![Screenshot from 2022-08-31 18-05-45](https://user-images.githubusercontent.com/369018/187800406-3c1525c8-94f5-4fdf-a8ba-e2c815d148f9.png)

ModulesUnraveled commented on components/03-organisms/menu/breadcrumbs/yds-breadcrumbs.js:
```diff
@@ -2,15 +2,229 @@ Drupal.behaviors.breadcrumbs = {
+     */
+    function mouseNav(direction) {
+      // If right.
+      if (direction === 'right') {
```
I like this, but the linter isn't happy
![Screen Shot 2022-09-02 at 4 36 31 PM](https://user-images.githubusercontent.com/1663810/188238544-c6052705-15fc-404a-9a20-072a542a4a57.png)

ModulesUnraveled replied:
This seems to work, and still have a happy linter
```
    function mouseNav(direction) {
      breadcrumbsMenu.scrollLeft =
        direction === 'right' ? getFirstVisible() : getLastHidden();
    }
```

callinmullaney replied:
nice!

ModulesUnraveled commented on components/03-organisms/menu/breadcrumbs/yds-breadcrumbs.js:
```diff
@@ -2,15 +2,229 @@ Drupal.behaviors.breadcrumbs = {
+          .getBoundingClientRect().right,
+      );
+
+      if (firstBreadcrumbsLeft < breadcrumbsLeft) {
```
I tried to refactor this, but because there's nested logic, the switch statement was equally spaghetti... and it had some weird bugs (the arrows were showing and hiding in the middle of scrolling). I probably didn't do it right, but it seems that nested switch statements aren't really preferred either.

So, I'm going to leave it as is for now.

callinmullaney replied:
word 👍

ModulesUnraveled commented on components/03-organisms/menu/breadcrumbs/yds-breadcrumbs.js:
```diff
@@ -2,15 +2,229 @@ Drupal.behaviors.breadcrumbs = {
+        }
+      });
+
+      return hiddenItems[hiddenItems.length - 1].offsetLeft - controlsWidth;
```
I think the only way to prevent this would be to set some sort of "disabled" flag on the arrow until the "transition" is done (the transition being the time it takes for the arrow to disappear, if needed) ... I think that's adding a lot of complexity for not a lot of benefit. Even if the JS error throws, it doesn't really hurt anything.

I'll add a todo to the code, but I'm not sure this will ever be a priority.

callinmullaney replied:
@ModulesUnraveled I hear ya and will approve because it's likely an edge case. That being said, a JS error this high up on a page may result in JS not firing properly further down the page for another component. Something to consider.

ModulesUnraveled commented:
@ntay90 I'd like to talk this through with you next week. I wonder if we can make the current page focusable like Callin mentioned, and still maintain a11y expectations... I don't know for sure, but I think talking it through might help.

ModulesUnraveled commented:
@ntay90 @callinmullaney I refactored this to use `aria-current="page"` on the last item (and made it a link) so that it is focusable, and maintained the visual difference as well. It's ready for your re-reviews!

ModulesUnraveled commented:
@callinmullaney 
1. The mobile collapsed and expanded versions are built how they're designed. I think we should leave them for now, and see if any feedback comes in from user tests or general feedback.
2. Like we discussed on zoom, this would require even more custom JS, so let's run it through user testing, and see if that's worth the effort before we write it.
3. I fixed this with Nick - we changed the last item to a link, and used `aria-current` to indicate that it's the current page for screen reader users.

callinmullaney reviewed (approved):
Per our Zoom convo we are throwing over the fence the UX comments I left previously. Other comments have been addressed and code looks good :shipit:
